File: src/go/matrixCalculator/matrixCalculator.go
package matrixCalculator

import (
	"errors"
)

import results "../results"

func createEmptyMatrix(side int) [][]int {
	result := [][]int{}
	for rowIndex := 0; rowIndex < side; rowIndex++ {
		row := []int{}
		for columnIndex := 0; columnIndex < side; columnIndex++ {
			row = append(row, 0)
		}
		result = append(result, row)
	}
	return result
}

type currentHolder struct {
	Row         int
	Column      int
	side        int
	StartRow    int
	StartColumn int
}

func (current *currentHolder) actual(value int) int {
	if value < 0 {
		return value + current.side
	} else if value >= current.side {
		return value - current.side
	}

	return value
}

func (current *currentHolder) actualRow() int {
	return current.actual(current.Row)
}

func (current *currentHolder) actualColumn() int {
	return current.actual(current.Column)
}

// Calculate calculates the matrix
func Calculate(side int) (results.MatrixResult, error) {
	if side%2 != 1 {
		return results.MatrixResult{}, errors.New("Side needs to be an odd number")
	}

	expectedSum := side * (side*side + 1) / 2
	matrix := createEmptyMatrix(side)
	current := currentHolder{
		-(side - 1) / 2,
		(side - 1) / 2,
		side,
		-(side - 1) / 2,
		(side - 1) / 2}

	for count := 1; count <= side*side; count++ {
		matrix[current.actualRow()][current.actualColumn()] = count

		if count%side != 0 {
			current.Row++
			current.Column++
		} else {
			current.StartRow++
			current.Row = current.StartRow
			current.StartColumn--
			current.Column = current.StartColumn
		}
	}

	return results.MatrixResult{matrix, expectedSum, side, true}, nil
}

File: src/erl/matrixCalculator.erl
-module(matrixCalculator).
-export([calculate/1]).

-include("results.hrl").

-record(currentHolder,{row, column, startRow, startColumn, side}).

createEmptyMatrix(Side) ->
    Matrix = array:new(Side),
    array:map(fun(_,_) -> array:new(Side) end, Matrix).

setMatrix(RowIndex, ColumnIndex, Value, Matrix) ->
    OriginalRow = array:get(RowIndex, Matrix),
    ChangedRow = array:set(ColumnIndex, Value, OriginalRow),
    array:set(RowIndex, ChangedRow, Matrix).


getMatrix(RowIndex, ColumnIndex, Matrix) ->
    Row = array:get(RowIndex, Matrix),
    array:get(ColumnIndex, Row).
    
actual(Value, Side) ->
    if 
        Value < 0 -> Value + Side;
        Value >= Side -> Value - Side;
        true -> Value
    end.

actualRow(Current) ->
    actual(Current#currentHolder.row, Current#currentHolder.side).

actualColumn(Current) ->
    actual(Current#currentHolder.column, Current#currentHolder.side).

getNext(Current, Count) ->
    if 
        Count rem Current#currentHolder.side /= 0 ->
            Current#currentHolder{row=Current#currentHolder.row + 1, column=Current#currentHolder.column + 1};
        true ->
            StartRow = Current#currentHolder.startRow + 1,
            StartColumn = Current#currentHolder.startColumn - 1,
            Current#currentHolder{row=StartRow, column=StartColumn, startRow=StartRow, startColumn=StartColumn}
    end.

testResult(Matrix, ExpectedSum, Side) ->


calculate(Side) ->
    if
        Side rem 2 == 0 -> throw("Side needs to be an odd number.");
        true -> ok
    end,

    ExpectedSum = round(Side * (Side * Side + 1) / 2),
    put(matrix, createEmptyMatrix(Side)),
    StartRow = round(-(Side - 1) / 2),
    StartColumn = round((Side - 1) / 2),
    put(current, #currentHolder{row=StartRow, column=StartColumn, startRow=StartRow, startColumn=StartColumn, side=Side}),

    lists:foreach(fun (Count) -> 
        Current = get(current),
        Matrix = get(matrix),
        put(matrix, setMatrix(actualRow(Current), actualColumn(Current), Count, Matrix)),
        put(current, getNext(Current, Count))
    end, lists:seq(1, Side * Side)),

    #matrixResult{success=true, matrix=get(matrix), sum=ExpectedSum, side=Side}.